src/main/java/apoc/refactor/GraphRefactoring.java

```diff
@@ -416,65 +442,81 @@ public void categorize(
 
         copiedKeys.remove(targetKey); // Just to be sure
 
-        // Create batches of nodes
-        List<Node> batch = null;
-        List<Future<Void>> futures = new ArrayList<>();
-        try (Transaction tx = db.beginTx()) {
-            for (Node node : db.getAllNodes()) {
-                if (batch == null) {
-                    batch = new ArrayList<>((int) batchSize);
-                }
-                batch.add(node);
-                if (batch.size() == batchSize) {
-                    futures.add(categorizeNodes(batch, sourceKey, relationshipType, outgoing, label, targetKey, copiedKeys));
-                    batch = null;
-                }
-            }
-            if (batch != null) {
-                futures.add(categorizeNodes(batch, sourceKey, relationshipType, outgoing, label, targetKey, copiedKeys));
-            }
+        final String[] copiedKeysArray = copiedKeys.toArray(new String[copiedKeys.size()]);
+
+        // We group the nodes by the sourceKey property value
+        Map<Object, Collection<List<Node>>> catChunksMap = getNodeCategoryMap(sourceKey, batchSize);
```

Comment:
Does this create a Map containing all the nodes that are being refactored? It seems like that would take a lot of memory if that's the case...

Reply:
Yep, you're right but how can we group the dataset by the categorization nodes without materializing the dataset?

Please look at the comment here:
https://github.com/neo4j-contrib/neo4j-apoc-procedures/issues/826#issuecomment-396156290